src/lib/y2users/linux/writer.rb

```diff
@@ -158,6 +165,68 @@ def add_user(user, issues)
         log.error("Error creating user '#{user.name}' - #{e.message}")
       end
 
+      # Command for creating new users
+      GROUPADD = "/usr/sbin/useradd".freeze
+      private_constant :USERADD
+
+      # Executes the command for creating the group
+      #
+      # @param group [Y2User::Group] the group to be created on the system
+      # @param issues [Y2Issues::List] a collection for adding an issue if something goes wrong
+      def add_group(group, issues)
+        args = [GROUPADD]
+        args << "--gid" << group.gid if group.gid
+        # TODO: system groups?
+        Yast::Execute.on_target!(args)
+      rescue Cheetah::ExecutionFailed => e
+        issues << Y2Issues::Issue.new(
+          format(_("The user '%{username}' could not be created"), username: user.name)
+        )
+        log.error("Error creating user '#{user.name}' - #{e.message}")
+      end
+
+      USERMOD_ATTRS = [:home, :shell, :gecos].freeze
+
+      def modify(new_user, old_user, issues)
+        change_password(new_user, issues) if new_user.password != old_user.password
+
+        usermod_changes = USERMOD_ATTRS.any? do |attr|
+          !new_user.public_send(attr).nil? &&
+            (new_user.public_send(attr) != old_user.public_send(attr))
+        end
+        # usermod also manage suplimentary groups, so compare also them
+        usermod_changes ||= different_groups?(new_user, old_user)
+        usermod_modify(new_user, old_user, issues) if usermod_changes
+      end
+
+      def different_groups?(lhu, rhu)
+        lhu.groups(with_primary: false).sort != rhu.groups(with_primary: false).sort
+      end
+
+      USERMOD = "/usr/sbin/usermod".freeze
+      private_constant :USERMOD
+      def usermod_modify(new_user, old_user, issues)
```

Comment:
Actually, it does not matter if you run the command with the same previous values, right? So, I'd simplify this method just doing something similar that it was done in useradd_options: include all options with the new_user values, reject those nil values and run the command.


```ruby
opts = {
  "--home"    => new_user.home,
  "--shell"   => new_user.shell,
  "--comment" => new_user.gecos.join(","),
  "--groups"  => new_users.groups(with_primary: false).join(",")
}

opts = opts.compact.flatten
opts << "--move-home" if opts.include?("--home")
opts << new_user.name

Yast::Execute.on_target!(USERMOD, opts)
```

Reply:
well, I am not sure if it matters. E.g. if home is same and option move-home is used if it do something or not. So to be sure I use approach which pass it only when it really is needed.

Reply:
I see. Anyway, I did some test before commenting and it looks to me that `usermod` is smart enough to only perform real changes. That said, its man page says nothing to that regard.

Reply:
>  E.g. if home is same and option move-home is used if it do something or not

Just out of curiosity, I have been taking a look to the [`usermod` source code](https://github.com/shadow-maint/shadow/blob/master/src/usermod.c) and checked that it prevents trying to [move_home](https://github.com/shadow-maint/shadow/blob/ca1d912e6b38c4e065592d4cc61d8365e4210882/src/usermod.c#L2398-L2400) by [overriding the flag](https://github.com/shadow-maint/shadow/blob/3ac8d97825f01055974039dd29eb1fb59c380b23/src/usermod.c#L1423-L1427) when both, given directory is NULL or it is the same.

Reply:
Also checked that [we're not changing](https://build.opensuse.org/package/show/SUSE:SLE-15-SP3:GA/shadow) its behavior ;)